# NumPy — Complete Hands-On Guide
## Numerical Python: The Foundation of All Data Science

---

**What is NumPy?**
NumPy (Numerical Python) is the core library for fast numerical computing in Python.
Almost every data-science and ML library (Pandas, Scikit-learn, TensorFlow, PyTorch) is built on top of it.

**Why NumPy over plain Python lists?**
| Feature | Python List | NumPy Array |
|---|---|---|
| Speed | Slow (interpreted loops) | Fast (vectorised C code) |
| Memory | High | Low (contiguous block) |
| Math operations | Manual loops | One-line operations |
| Multi-dimensional | Nested lists | Native n-D arrays |
| Data types | Mixed | Homogeneous (faster) |

**Topics covered in this notebook:**
1. Creating Arrays
2. Array Properties
3. Indexing & Slicing
4. Array Math Operations
5. Statistical Functions
6. Reshaping & Manipulation
7. Boolean Masking
8. Broadcasting
9. Sorting Arrays
10. Stacking & Splitting
11. Random Numbers
12. Linear Algebra Basics
13. Practice Assignment

---
> Run each cell with **Shift + Enter**


---
## 1. Creating Arrays

NumPy gives you many ways to build an array — from data, from ranges, or from special patterns.


In [ ]:
import numpy as np

# ── From Python lists ──
arr_1d = np.array([10, 20, 30, 40, 50])
arr_2d = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])
arr_3d = np.array([[[1, 2], [3, 4]],
                   [[5, 6], [7, 8]]])

print("1D array:", arr_1d)
print("\n2D array (3×3):\n", arr_2d)
print("\n3D array (2×2×2):\n", arr_3d)


In [41]:
import numpy as np

# ── Special-value arrays ──
zeros   = np.zeros((3, 4))          # all zeros
ones    = np.ones((2, 3))           # all ones
full    = np.full((3, 3), 7)        # filled with 7
eye     = np.eye(4)                 # identity matrix (4×4)
empty   = np.empty((2, 2))          # uninitialised (whatever's in memory)

print("Zeros (3×4):\n", zeros)
print("\nOnes (2×3):\n", ones)
print("\nFull of 7s (3×3):\n", full)
print("\nIdentity (4×4):\n", eye)


Zeros (3×4):
 [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]

Ones (2×3):
 [[1. 1. 1.]
 [1. 1. 1.]]

Full of 7s (3×3):
 [[7 7 7]
 [7 7 7]
 [7 7 7]]

Identity (4×4):
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


In [3]:
import numpy as np

# ── Range-based arrays ──
rng1 = np.arange(0, 20, 3)              # like Python range — start, stop, step
rng2 = np.linspace(0, 1, 6)            # 6 evenly spaced values between 0 and 1
rng3 = np.linspace(0, 100, 5)          # 5 values between 0 and 100
rng4 = np.linspace(5, 50, 10)

print("arange(0,20,3)       :", rng1)
print("linspace(0,1,6)      :", rng2.round(2))
print("linspace(0,100,5)    :", rng3)
print("np.linspace(5, 50, 10): ", rng4)

# ── Specifying data type ──
int_arr   = np.array([1.9, 2.7, 3.1], dtype=np.int32)   # truncates to int
float_arr = np.array([1, 2, 3],       dtype=np.float64)
bool_arr  = np.array([0, 1, 0, 1, 1], dtype=bool)
rng4_int = np.array(rng4, dtype=np.int32)

print("\nWith dtype=int32 :", int_arr)
print("With dtype=float64:", float_arr)
print("With dtype=bool   :", bool_arr)
print("rng4_int32: ", rng4_int)


arange(0,20,3)       : [ 0  3  6  9 12 15 18]
linspace(0,1,6)      : [0.  0.2 0.4 0.6 0.8 1. ]
linspace(0,100,5)    : [  0.  25.  50.  75. 100.]
np.linspace(5, 50, 10):  [ 5. 10. 15. 20. 25. 30. 35. 40. 45. 50.]

With dtype=int32 : [1 2 3]
With dtype=float64: [1. 2. 3.]
With dtype=bool   : [False  True False  True  True]
rng4_int32:  [ 5 10 15 20 25 30 35 40 45 50]


---
## 2. Array Properties

Always inspect an array before working with it — shape, size, and dtype are critical.


In [31]:
import numpy as np

arr = np.array([[1.5, 2.3, 3.7],
                [4.1, 5.9, 6.0]])

print("Array:\n", arr)
print("\n─── Properties ───")
print("arr.shape    :", arr.shape)       # (rows, cols)
print("arr.ndim     :", arr.ndim)        # number of dimensions
print("arr.size     :", arr.size)        # total elements
print("arr.dtype    :", arr.dtype)       # element data type
print("arr.itemsize :", arr.itemsize, "bytes")  # bytes per element
print("arr.nbytes   :", arr.nbytes, "bytes")    # total memory used

# ── Type conversion ──
print("\n─── Type Conversion ───")
print("astype(int)  :", arr.astype(int))
print("astype(str)  :", arr.astype(str))

print("*****TESTING*****")
np_arr = np.array([[[1,2,3],[4,5,6],[0,0,0]], [[7,8,9],[10,11,12],[0,0,0]], [[13,14,15], [16,17,20],[0,0,0]]], dtype=np.int32)

print(f"{np_arr.shape= }") # 3 layers, 3 rows, 3 columns
print(f"{np_arr.ndim= }") # tells the array dimention
print(f"{np_arr.size= }") # tells total number of records in an array
print(f"{np_arr.dtype= }") # print the data type of the array
print(f"{np_arr.itemsize= }") # it print byte per elements
print(f"{np_arr.nbytes= }") #total memory used by the np_arr


Array:
 [[1.5 2.3 3.7]
 [4.1 5.9 6. ]]

─── Properties ───
arr.shape    : (2, 3)
arr.ndim     : 2
arr.size     : 6
arr.dtype    : float64
arr.itemsize : 8 bytes
arr.nbytes   : 48 bytes

─── Type Conversion ───
astype(int)  : [[1 2 3]
 [4 5 6]]
astype(str)  : [['1.5' '2.3' '3.7']
 ['4.1' '5.9' '6.0']]
*****TESTING*****
np_arr.shape= (3, 3, 3)
np_arr.ndim= 3
np_arr.size= 27
np_arr.dtype= dtype('int32')
np_arr.itemsize= 4
np_arr.nbytes= 108


In [32]:
import numpy as np

# ── Copy vs View (very important!) ──
original = np.array([1, 2, 3, 4, 5])

view   = original[1:4]     # a VIEW — shares memory with original
print(f"{view=}")
copy   = original[1:4].copy()  # a COPY — independent

view[0]  = 999   # modifies original too!
copy[0]  = 888   # does NOT modify original

print("After view[0] = 999:")
print("  original :", original)   # [1, 999, 3, 4, 5]
print("  view     :", view)       # [999, 3, 4]

print("\nAfter copy[0] = 888:")
print("  original :", original)   # unchanged
print("  copy     :", copy)       # [888, 3, 4]

# Check if an array is a view or a copy
print("\nview.base is original :", view.base is original)   # True
print("copy.base is None     :", copy.base is None)         # True


view=array([2, 3, 4])
After view[0] = 999:
  original : [  1 999   3   4   5]
  view     : [999   3   4]

After copy[0] = 888:
  original : [  1 999   3   4   5]
  copy     : [888   3   4]

view.base is original : True
copy.base is None     : True


---
## 3. Indexing & Slicing

Access specific elements, rows, columns, sub-arrays, or arbitrary indices.


In [33]:
import numpy as np

arr = np.array([[10, 20, 30, 40],
                [50, 60, 70, 80],
                [90,100,110,120]])

print("Array:\n", arr)

# ── Single element [row, col] ──
print("\narr[0, 0]  =", arr[0, 0])    # first element
print("arr[1, 2]  =", arr[1, 2])       # row 1, col 2 → 70
print("arr[-1,-1] =", arr[-1, -1])     # last element → 120

# ── Slicing [row_start:row_stop, col_start:col_stop] ──
print("\nFirst 2 rows        :\n", arr[:2, :])
print("Last 2 columns      :\n", arr[:, 2:])
print("Middle 2×2 block    :\n", arr[0:2, 1:3])
print("Every other column  :\n", arr[:, ::2])

# ── Full row / column ──
print("\nRow 2    :", arr[2, :])
print("Column 1 :", arr[:, 1])


Array:
 [[ 10  20  30  40]
 [ 50  60  70  80]
 [ 90 100 110 120]]

arr[0, 0]  = 10
arr[1, 2]  = 70
arr[-1,-1] = 120

First 2 rows        :
 [[10 20 30 40]
 [50 60 70 80]]
Last 2 columns      :
 [[ 30  40]
 [ 70  80]
 [110 120]]
Middle 2×2 block    :
 [[20 30]
 [60 70]]
Every other column  :
 [[ 10  30]
 [ 50  70]
 [ 90 110]]

Row 2    : [ 90 100 110 120]
Column 1 : [ 20  60 100]


In [34]:
import numpy as np

arr = np.array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100])

# ── 1D slicing ──
print("arr[2:6]   :", arr[2:6])     # elements at index 2,3,4,5
print("arr[:5]    :", arr[:5])      # first 5
print("arr[-3:]   :", arr[-3:])     # last 3
print("arr[::2]   :", arr[::2])     # every other element
print("arr[::-1]  :", arr[::-1])    # reversed

# ── Fancy indexing — use an array of indices ──
idx = np.array([0, 2, 4, 9])
print("\nFancy index [0,2,4,9]:", arr[idx])

# ── 2D fancy indexing ──
matrix = np.array([[1,  2,  3],
                   [4,  5,  6],
                   [7,  8,  9],
                   [10,11, 12]])
rows = [0, 2, 3]
cols = [1, 0, 2]
print("\nPick (row0,col1), (row2,col0), (row3,col2):", matrix[rows, cols])


arr[2:6]   : [30 40 50 60]
arr[:5]    : [10 20 30 40 50]
arr[-3:]   : [ 80  90 100]
arr[::2]   : [10 30 50 70 90]
arr[::-1]  : [100  90  80  70  60  50  40  30  20  10]

Fancy index [0,2,4,9]: [ 10  30  50 100]

Pick (row0,col1), (row2,col0), (row3,col2): [ 2  7 12]


---
## 4. Array Math Operations

NumPy performs element-wise math without any loops — this is called **vectorisation**.


In [35]:
import numpy as np

a = np.array([10, 20, 30, 40, 50])
b = np.array([ 1,  2,  3,  4,  5])

print("a =", a)
print("b =", b)

# ── Element-wise arithmetic ──
print("\na + b   =", a + b)
print("a - b   =", a - b)
print("a * b   =", a * b)
print("a / b   =", a / b)
print("a // b  =", a // b)    # floor division
print("a ** 2  =", a ** 2)    # squaring
print("a % 3   =", a % 3)     # modulus

# ── Scalar operations ──
print("\na + 100 =", a + 100)
print("a * 0.5 =", a * 0.5)
print("a > 25  =", a > 25)    # comparison returns boolean array


a = [10 20 30 40 50]
b = [1 2 3 4 5]

a + b   = [11 22 33 44 55]
a - b   = [ 9 18 27 36 45]
a * b   = [ 10  40  90 160 250]
a / b   = [10. 10. 10. 10. 10.]
a // b  = [10 10 10 10 10]
a ** 2  = [ 100  400  900 1600 2500]
a % 3   = [1 2 0 1 2]

a + 100 = [110 120 130 140 150]
a * 0.5 = [ 5. 10. 15. 20. 25.]
a > 25  = [False False  True  True  True]


In [ ]:
import numpy as np

a = np.array([10, 20, 30, 40, 50])

# ── Universal functions (ufuncs) ──
print("np.sqrt(a)      =", np.sqrt(a).round(2))       # square root of each element
print("np.log(a)       =", np.log(a).round(2))         # natural log (base e) of each element
print("np.log10(a)     =", np.log10(a).round(2))       # log base 10 of each element
print("np.exp([1,2,3]) =", np.exp([1, 2, 3]).round(2)) # e raised to the power of each element
print("np.abs([-3,1,-5])=", np.abs([-3, 1, -5]))       # absolute value, removes negative sign
print("np.floor([1.9]) =", np.floor([1.9, 2.1, -1.5])) # rounds down to nearest integer
print("np.ceil([1.1])  =", np.ceil([1.1, 2.9, -1.1]))  # rounds up to nearest integer
print("np.round([1.567],2)=", np.round([1.567, 2.345], 2)) # rounds to given decimal places

# ── Dot product & matrix multiplication ──
v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])
print("\nDot product v1·v2:", np.dot(v1, v2))   # sum of element-wise multiplication: 1*4 + 2*5 + 3*6 = 32

---
## 5. Statistical Functions

NumPy has built-in stats — mean, median, std, percentiles — with **axis** control.

> **axis=0** → operate *down* rows (result is one value per column)
> **axis=1** → operate *across* columns (result is one value per row)


In [ ]:
import numpy as np

# Student exam scores: rows = students, cols = subjects
scores = np.array([[85, 92, 78, 90],
                   [70, 65, 88, 72],
                   [95, 88, 91, 97],
                   [60, 75, 82, 68]])

print("Score matrix (4 students × 4 subjects):\n", scores)

print("\n─── Overall ───")
print("Dimention: ", scores.ndim)                                # number of dimensions of the array
print("Min    :", np.min(scores))                                # smallest value in the entire array
print("Max    :", np.max(scores))                                # largest value in the entire array
print("Mean   :", np.mean(scores))                               # average of all elements (sum / total count)
print("Median :", np.median(scores))                             # middle value when all elements are sorted
print("Std    :", np.std(scores).round(2))                       # standard deviation — how spread out values are from the mean
print("Sum    :", np.sum(scores))                                # total of all elements added together
print("Var    :", np.var(scores).round(2))                       # variance — square of standard deviation

print("\n─── Per Student (axis=1) ───")
print("Mean per student :", np.mean(scores, axis=1))             # average score of each student across all subjects
print("Total per student:", np.sum(scores, axis=1))              # total score of each student across all subjects

print("\n─── Per Subject (axis=0) ───")
print("Mean per subject :", np.mean(scores, axis=0))             # average score per subject across all students
print("Max per subject  :", np.max(scores, axis=0))              # highest score per subject across all students

print("\n─── Percentiles ───")
print("25th percentile:", np.percentile(scores, 25))             # value below which 25% of data falls
print("50th (median)  :", np.percentile(scores, 50))             # value below which 50% of data falls (same as median)
print("75th percentile:", np.percentile(scores, 75))             # value below which 75% of data falls

In [42]:
import numpy as np

data = np.array([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5])

# ── Cumulative operations ──
print("data       :", data)
print("cumsum     :", np.cumsum(data))    # running total — each element is sum of all previous elements
print("cumprod    :", np.cumprod(data))   # running product — each element is product of all previous elements

# ── argmin / argmax — index of the min/max ──
print("\nargmin:", np.argmin(data), "→ value:", data[np.argmin(data)])   # index position of the smallest value
print("argmax:", np.argmax(data), "→ value:", data[np.argmax(data)])   # index position of the largest value

# ── Correlation and covariance ──
x = np.array([1, 2, 3, 4, 5])
y = np.array([5, 4, 3, 2, 1])  # perfectly negatively correlated
print("\ncorrcoef(x, y):\n", np.corrcoef(x, y))   # correlation matrix — measures linear relationship (-1 to +1)
print("cov(x, y)      :\n", np.cov(x, y))         # covariance matrix — how two variables change together

data       : [3 1 4 1 5 9 2 6 5 3 5]
cumsum     : [ 3  4  8  9 14 23 25 31 36 39 44]
cumprod    : [     3      3     12     12     60    540   1080   6480  32400  97200
 486000]

argmin: 1 → value: 1
argmax: 5 → value: 9

corrcoef(x, y):
 [[ 1. -1.]
 [-1.  1.]]
cov(x, y)      :
 [[ 2.5 -2.5]
 [-2.5  2.5]]


---
## 6. Reshaping & Manipulation

Change the shape of an array without changing its data.
This is essential for preparing data for ML models.


In [43]:
import numpy as np

arr = np.arange(1, 25)   # 24 elements
print("Original (1D, 24 elements):", arr)

# ── reshape — total elements must stay the same ──
r4x6   = arr.reshape(4, 6)
r3x8   = arr.reshape(3, 8)
r2x3x4 = arr.reshape(2, 3, 4)   # 3D!

print("\nReshape (4×6):\n",   r4x6)
print("\nReshape (3×8):\n",   r3x8)
print("\nReshape (2×3×4) — 3D:\n", r2x3x4)

# ── -1 lets NumPy infer one dimension ──
auto = arr.reshape(-1, 6)   # NumPy figures out rows = 4
print("\nreshape(-1, 6):", auto.shape)


Original (1D, 24 elements): [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]

Reshape (4×6):
 [[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]]

Reshape (3×8):
 [[ 1  2  3  4  5  6  7  8]
 [ 9 10 11 12 13 14 15 16]
 [17 18 19 20 21 22 23 24]]

Reshape (2×3×4) — 3D:
 [[[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]

 [[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]]

reshape(-1, 6): (4, 6)


In [44]:
import numpy as np

matrix = np.array([[1, 2, 3],
                   [4, 5, 6]])
print("Original (2×3):\n", matrix)

# ── Flatten / Ravel ──
print("\nflatten() (copy) :", matrix.flatten())
print("ravel()   (view) :", matrix.ravel())

# ── Transpose ──
print("\nTranspose (T) shape:", matrix.T.shape)
print(matrix.T)

# ── expand_dims — add an axis (common for ML) ──
arr1d = np.array([1, 2, 3])             # shape (3,)
col   = np.expand_dims(arr1d, axis=1)   # shape (3,1) — column vector
row   = np.expand_dims(arr1d, axis=0)   # shape (1,3) — row vector
print("\noriginal shape:", arr1d.shape)
print("expand_dims(axis=1) shape:", col.shape)
print("expand_dims(axis=0) shape:", row.shape)

# ── squeeze — remove dimensions of size 1 ──
big = np.array([[[1, 2, 3]]])   # shape (1,1,3)
print("\nbefore squeeze:", big.shape)
print("after  squeeze:", np.squeeze(big).shape)


Original (2×3):
 [[1 2 3]
 [4 5 6]]

flatten() (copy) : [1 2 3 4 5 6]
ravel()   (view) : [1 2 3 4 5 6]

Transpose (T) shape: (3, 2)
[[1 4]
 [2 5]
 [3 6]]

original shape: (3,)
expand_dims(axis=1) shape: (3, 1)
expand_dims(axis=0) shape: (1, 3)

before squeeze: (1, 1, 3)
after  squeeze: (3,)


---
## 7. Boolean Masking (Filtering)

Select, count, or modify elements based on conditions — without any explicit loops.


In [45]:
import numpy as np

temps = np.array([22, 35, 18, 40, 15, 38, 27, 12, 33, 29])
print("Temperatures:", temps)

# ── Build masks ──
hot_mask  = temps > 30
cold_mask = temps < 20
mild_mask = (temps >= 20) & (temps <= 30)

print("\nHot  mask (>30) :", hot_mask)
print("Hot  days       :", temps[hot_mask])
print("Cold days (<20) :", temps[cold_mask])
print("Mild days (20-30):", temps[mild_mask])

# ── Count and percentage ──
print("\nHot  day count      :", np.sum(hot_mask))
print("Cold day count      :", np.sum(cold_mask))
print("Hot  day percentage :", np.mean(hot_mask) * 100, "%")

# ── Combine conditions with & (AND) | (OR) ~ (NOT) ──
extreme = (temps > 35) | (temps < 15)
print("\nExtreme temps (>35 or <15):", temps[extreme])
not_hot = temps[~hot_mask]
print("Not hot temps          :", not_hot)


Temperatures: [22 35 18 40 15 38 27 12 33 29]

Hot  mask (>30) : [False  True False  True False  True False False  True False]
Hot  days       : [35 40 38 33]
Cold days (<20) : [18 15 12]
Mild days (20-30): [22 27 29]

Hot  day count      : 4
Cold day count      : 3
Hot  day percentage : 40.0 %

Extreme temps (>35 or <15): [40 38 12]
Not hot temps          : [22 18 15 27 12 29]


In [ ]:
import numpy as np

temps = np.array([22, 35, 18, 40, 15, 38, 27, 12, 33, 29])

# ── np.where — element-wise if/else ──
labels = np.where(temps > 30, "HOT",
         np.where(temps < 20, "COLD", "MILD"))
print("Labels:", labels)

# ── np.select — multiple conditions ──
conditions = [temps < 15, temps < 25, temps <= 35, temps > 35]
choices    = ["FREEZING", "COLD", "WARM", "SCORCHING"]
category   = np.select(conditions, choices, default="MODERATE")
print("Category:", category)

# ── Modify in-place using mask ──
temps_copy = temps.copy()
temps_copy[temps_copy > 35] = 35    # cap at 35
print("\nCapped at 35:", temps_copy)

# ── np.clip — clamp values between min and max ──
clipped = np.clip(temps, 15, 35)    # same result, one line
print("np.clip(15,35):", clipped)


---
## 8. Broadcasting

Apply operations between arrays of **different shapes**.
NumPy automatically "stretches" smaller arrays to match larger ones.

**Broadcasting Rules:**
1. If arrays have different numbers of dimensions, the smaller shape is padded with 1s on the left.
2. Dimensions of size 1 are stretched to match the other array.
3. If shapes are incompatible and neither is 1, NumPy raises an error.


In [46]:
import numpy as np

# ── Simple broadcasting examples ──
arr    = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])   # shape (3, 3)

row_v  = np.array([10, 20, 30])  # shape (3,)  — broadcast across rows
col_v  = np.array([[1],[2],[3]]) # shape (3,1) — broadcast across columns

print("Array:\n", arr)
print("\n+ row vector [10,20,30]:\n", arr + row_v)
print("\n+ col vector [[1],[2],[3]]:\n", arr + col_v)
print("\nrow + col:\n", row_v + col_v)


Array:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]

+ row vector [10,20,30]:
 [[11 22 33]
 [14 25 36]
 [17 28 39]]

+ col vector [[1],[2],[3]]:
 [[ 2  3  4]
 [ 6  7  8]
 [10 11 12]]

row + col:
 [[11 21 31]
 [12 22 32]
 [13 23 33]]


In [ ]:
import numpy as np

# ── Real-world: Z-score normalisation ──
# (subtract column mean, divide by column std)
scores = np.array([[80, 90, 70],
                   [60, 75, 85],
                   [95, 65, 88]])

mean_col = scores.mean(axis=0)   # shape (3,) — mean of each subject
std_col  = scores.std(axis=0)    # shape (3,)

# Broadcasting: (3,3) - (3,) → NumPy broadcasts (3,) across all rows
z_scores = (scores - mean_col) / std_col

print("Original scores:\n", scores)
print("\nMean per subject :", mean_col.round(1))
print("Std  per subject :", std_col.round(2))
print("\nZ-score normalised:\n", z_scores.round(2))
print("\nColumn means after normalisation:", z_scores.mean(axis=0).round(10))


---
## 9. Sorting Arrays

Sort values, find sorted indices, and search in sorted arrays.


In [ ]:
import numpy as np

arr = np.array([3, 1, 4, 1, 5, 9, 2, 6, 5, 3])
print("Original:", arr)

# ── np.sort — returns a SORTED COPY ──
print("Ascending  :", np.sort(arr))
print("Descending :", np.sort(arr)[::-1])

# ── arr.sort() — sorts IN PLACE ──
arr_copy = arr.copy()
arr_copy.sort()
print("In-place sorted:", arr_copy)

# ── argsort — returns indices that would sort the array ──
arr2 = np.array([40, 10, 30, 20])
idx  = np.argsort(arr2)
print("\nargsort indices:", idx)
print("Sorted via argsort:", arr2[idx])

# ── 2D sorting ──
matrix = np.array([[3, 1, 4],
                   [9, 2, 6],
                   [5, 8, 7]])
print("\nOriginal matrix:\n", matrix)
print("Sorted axis=1 (each row):\n", np.sort(matrix, axis=1))
print("Sorted axis=0 (each col):\n", np.sort(matrix, axis=0))


In [ ]:
import numpy as np

# ── np.unique — unique values and counts ──
data = np.array([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5])
unique_vals, counts = np.unique(data, return_counts=True)
print("Unique values:", unique_vals)
print("Counts       :", counts)

# ── np.searchsorted — find insertion index in sorted array ──
sorted_arr = np.array([10, 20, 30, 40, 50])
idx = np.searchsorted(sorted_arr, 35)
print("\nInsert 35 into", sorted_arr, "→ index", idx)

# ── np.partition — partial sort (faster than full sort for top-k) ──
arr = np.array([7, 3, 9, 1, 5, 6, 4, 8, 2])
print("\nOriginal:", arr)
print("Smallest 3:", np.partition(arr, 3)[:3])
print("Largest  3:", np.partition(arr, -3)[-3:])


---
## 10. Stacking & Splitting Arrays

Combine multiple arrays into one, or break one into pieces.


In [ ]:
import numpy as np

a = np.array([[1, 2], [3, 4]])
b = np.array([[5, 6], [7, 8]])

# ── Stacking ──
print("a:\n", a)
print("b:\n", b)

# vstack — stack vertically (add rows)
print("\nvstack (add rows):\n", np.vstack([a, b]))

# hstack — stack horizontally (add columns)
print("\nhstack (add cols):\n", np.hstack([a, b]))

# concatenate — explicit axis
print("\nconcatenate axis=0 (same as vstack):\n", np.concatenate([a, b], axis=0))
print("\nconcatenate axis=1 (same as hstack):\n", np.concatenate([a, b], axis=1))

# np.stack — adds a NEW axis
print("\nstack axis=0 shape:", np.stack([a, b], axis=0).shape)   # (2,2,2)
print("stack axis=2 shape:", np.stack([a, b], axis=2).shape)   # (2,2,2)


In [ ]:
import numpy as np

arr = np.arange(1, 25).reshape(4, 6)
print("Original (4×6):\n", arr)

# ── np.split — equal-size splits ──
# Split into 2 along axis=0 (rows)
top, bottom = np.split(arr, 2, axis=0)
print("\ntop half:\n", top)
print("bottom half:\n", bottom)

# Split into 3 equal columns
c1, c2, c3 = np.split(arr, 3, axis=1)
print("\ncol split 1:\n", c1)

# ── np.array_split — unequal splits ──
parts = np.array_split(arr, 3, axis=0)   # 4 rows ÷ 3 → [2,1,1]
for i, p in enumerate(parts):
    print(f"Part {i}: shape {p.shape}")


---
## 11. Random Numbers

NumPy's modern random API. Always set a **seed** for reproducible results.


In [ ]:
import numpy as np

# Modern API: np.random.default_rng(seed)
rng = np.random.default_rng(seed=42)

# ── Integers ──
dice    = rng.integers(1, 7, size=10)           # 10 dice rolls [1–6]
print("Dice rolls         :", dice)

# ── Floats (uniform 0 to 1) ──
uniform = rng.random(8)
print("Uniform [0,1)      :", uniform.round(3))

# ── Normal (Gaussian) distribution ──
heights = rng.normal(loc=170, scale=10, size=1000)   # mean=170cm, std=10cm
print(f"\nHeight simulation (n=1000):")
print(f"  Mean : {heights.mean():.1f} cm")
print(f"  Std  : {heights.std():.1f} cm")
print(f"  Min  : {heights.min():.1f} cm")
print(f"  Max  : {heights.max():.1f} cm")


In [ ]:
import numpy as np

rng = np.random.default_rng(seed=7)

# ── Choice — sample from an array ──
items = ['apple', 'banana', 'cherry', 'date', 'elderberry']
picked = rng.choice(items, size=3, replace=False)   # without replacement
print("Random pick (no repeat):", picked)

# ── Shuffle — in-place ──
arr = np.arange(1, 11)
rng.shuffle(arr)
print("Shuffled 1-10:", arr)

# ── Binomial distribution ──
coin_flips = rng.binomial(n=1, p=0.5, size=20)  # 20 fair coin flips
print("\nCoin flips (0=T, 1=H):", coin_flips)
print("Heads:", coin_flips.sum(), "/ 20")

# ── Exponential distribution ──
wait_times = rng.exponential(scale=5, size=8)   # mean wait = 5 mins
print("\nWait times (mins):", wait_times.round(2))


---
## 12. Linear Algebra Basics

NumPy's `linalg` module provides essential matrix operations used heavily in ML/DS.


In [ ]:
import numpy as np

A = np.array([[2, 1],
              [5, 3]])
B = np.array([[1, 2],
              [3, 4]])

# ── Matrix multiplication ──
print("A:\n", A)
print("B:\n", B)
print("\nA @ B (matrix multiply):\n", A @ B)          # preferred syntax
print("np.matmul(A, B):\n",            np.matmul(A, B))  # same result

# ── Dot product (for 1D vectors) ──
v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])
print("\nDot product:", np.dot(v1, v2))   # 1*4+2*5+3*6 = 32


In [ ]:
import numpy as np

A = np.array([[4.0, 7.0],
              [2.0, 6.0]])

# ── Determinant ──
det = np.linalg.det(A)
print("A:\n", A)
print("\nDeterminant:", det)   # 4*6 - 7*2 = 10

# ── Inverse ──
A_inv = np.linalg.inv(A)
print("\nInverse of A:\n", A_inv)

# ── Verify: A × A_inv should be identity matrix ──
product = A @ A_inv
print("\nA × A_inv (should be I):\n", product.round(10))

# ── Transpose ──
print("\nTranspose of A:\n", A.T)

# ── Norm (length/magnitude of a matrix or vector) ──
v = np.array([3, 4])
print("\n||[3,4]|| =", np.linalg.norm(v))   # sqrt(9+16) = 5


In [ ]:
import numpy as np

# ── Solve a system of linear equations ──
# 4x + 7y = 1
# 2x + 6y = 2
A = np.array([[4, 7],
              [2, 6]], dtype=float)
b = np.array([1, 2], dtype=float)

x = np.linalg.solve(A, b)
print("Solving: 4x + 7y = 1")
print("         2x + 6y = 2")
print("\nSolution:", x)
print("Verify A@x:", A @ x)   # should equal b

# ── Eigenvalues and Eigenvectors ──
M = np.array([[4, 1],
              [2, 3]])
eigenvalues, eigenvectors = np.linalg.eig(M)
print("\nMatrix M:\n", M)
print("Eigenvalues  :", eigenvalues)
print("Eigenvectors :\n", eigenvectors)


---
## Practice Assignment — 5 Questions

Try these on your own. Write your solutions in new cells below each question.

---

**Q1. Array Creation & Properties**
Create a 2D array of shape (5, 8) filled with random integers between 1 and 50 (seed=0).
Print: shape, dtype, total elements, mean, and the element at row 3, column 5.

---

**Q2. Reshaping & Slicing**
Create an array of integers from 1 to 36, reshape it to (6, 6).
- Extract the inner (4×4) block (rows 1–4, cols 1–4)
- Transpose the full matrix
- Flatten the full matrix to 1D

---

**Q3. Boolean Masking**
Given `sales = np.array([120, 85, 200, 45, 310, 95, 175, 60, 250, 130])`:
- Find all months where sales exceeded 150
- Count how many months had sales below 100
- Replace all values below 100 with exactly 100 (floor at 100)
- Classify each value as "LOW" (<100), "MID" (100–199), or "HIGH" (≥200) using `np.select`

---

**Q4. Broadcasting & Statistics**
Create a (4×5) matrix of random floats (seed=1), scaled to range 50–100.
- Compute mean and std per column
- Subtract the column mean from each element (z-score normalise)
- Verify that each column's mean is now ≈ 0

---

**Q5. Linear Algebra**
Create the matrix `M = [[3, 1], [2, 4]]`.
- Compute its determinant and inverse
- Verify M × M_inv ≈ Identity
- Solve the system: 3x + y = 9, 2x + 4y = 14
- Find the eigenvalues and eigenvectors

---
*Run your cells and compare results — there is only one correct answer for each!*
